### PR TITLE
[FEATURE] add quant algo `Learned Step Size Quantization`

### PR DESCRIPTION
**Pull request by FreakieHuang**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add LSQ QAT algorithm. Based on [humu's RFC initialization PR](https://github.com/open-mmlab/mmrazor/pull/342) 

## Modification

1. Add related quantization components
2. LSQ experiment result: R18 on cifar10
3. Comprehensive UT abstract classes. 
4. Add `custom trace`
5. Add `Learned Step Size Quantization`

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmrazor/models/observers/lsq_observer.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            self.min_val, self.max_val = torch._aminmax(y, 1)
+
+        return x
+
+    def calculate_qparams(self):
+        scale = 2 * self.tensor_norm / math.sqrt(self.quant_max)
+        zero_point = torch.zeros_like(self.tensor_norm)
+        sync_tensor(scale)
+        sync_tensor(zero_point)
+        if self.pot_scale:
```
Comment (humu789) [inline mmrazor/models/observers/lsq_observer.py:55]: `self.pot_scale` -> `self.is_pot_scale`

File: mmrazor/models/fake_quants/lsq.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    x = (x.round() - x).detach() + x
+    x = torch.clamp(x, quant_min, quant_max)
+    return (x - zero_point) * scale
+
+
+def grad_scale(t, scale):
+    return (t - (t * scale)).detach() + (t * scale)
+
+
+class FakeQuantizeLearnablePerchannelAffine(torch.autograd.Function):
```
Comment (humu789) [inline mmrazor/models/fake_quants/lsq.py:120]: need to take deployment into consideration later

**Review by humu789 — APPROVED** [review]

File: mmrazor/models/fake_quants/lsq.py (hunk 2 above)
Comment (FreakieHuang) [reply mmrazor/models/fake_quants/lsq.py:120]: Then update when need.

File: mmrazor/models/algorithms/quantization/base.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+    def prepare(self, mode='tensor'):
+        assert mode in ['tensor', 'loss', 'predict']
+        if mode == 'tensor':
+            graph = self.graph_tensor
+        elif mode == 'loss':
+            graph = self.graph_loss
+        else:
+            graph = self.graph_predict
+        self.architecture = self.quantizer.prepare(self.architecture, graph)
```
Comment (humu789) [inline mmrazor/models/algorithms/quantization/base.py:91]: the return graphmodule of `prepare` should assign to self.graph_mode

File: mmrazor/models/observers/minmax.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            x_dim = x.size()
+            new_axis_list = [i for i in range(len(x_dim))]
+            new_axis_list[self.ch_axis] = 0
+            new_axis_list[0] = self.ch_axis
+            y = x.permute(new_axis_list)
+            y = torch.flatten(y, start_dim=1)
+            min_val_cur, max_val_cur = torch._aminmax(y, 1)
+        min_val = torch.min(self.min_val, min_val_cur)
+        max_val = torch.max(self.max_val, max_val_cur)
+        self.min_val.copy_(min_val)
```
Comment (humu789) [inline mmrazor/models/observers/minmax.py:48]: when shapes of `min_val` and `self.min_val` are different, `copy_` will be error

**Pull request merged**